src/os/bluestore/aio.cc

```diff
@@ -71,54 +145,132 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
   return done;
 }
 
+#if defined(HAVE_LIBAIO)
 int aio_queue_t::get_next_completed(int timeout_ms, aio_t **paio, int max)
 {
-#if defined(HAVE_LIBAIO)
   io_event events[max];
-#elif defined(HAVE_POSIXAIO)
-  struct kevent events[max];
-#endif
   struct timespec t = {
     timeout_ms / 1000,
     (timeout_ms % 1000) * 1000 * 1000
   };
 
   int r = 0;
   do {
-#if defined(HAVE_LIBAIO)
     r = io_getevents(ctx, 1, max, events, &t);
-#elif defined(HAVE_POSIXAIO)
-    r = kevent(ctx, NULL, 0, events, max, &t);
-    if (r < 0)
-      r = -errno;
-#endif
   } while (r == -EINTR);
 
   for (int i=0; i<r; ++i) {
-#if defined(HAVE_LIBAIO)
     paio[i] = (aio_t *)events[i].obj;
     paio[i]->rval = events[i].res;
-#else
-    paio[i] = (aio_t*)events[i].udata;
-    if (paio[i]->n_aiocb == 1) {
-      paio[i]->rval = aio_return(&paio[i]->aio.aiocb);
-    } else {
-      // Emulate the return value of pwritev.  I can't find any documentation
-      // for what the value of io_event.res is supposed to be.  I'm going to
-      // assume that it's just like pwritev/preadv/pwrite/pread.
-      paio[i]->rval = 0;
-      for (int j = 0; j < paio[i]->n_aiocb; j++) {
-	int res = aio_return(&paio[i]->aio.aiocbp[j]);
-	if (res < 0) {
-	  paio[i]->rval = res;
-	  break;
-	} else {
-	  paio[i]->rval += res;
-	}
-      }
-      free(paio[i]->aio.aiocbp);
-    }
-#endif
   }
   return r;
 }
+#elif defined(HAVE_POSIXAIO)
+// return value: the number of completed AIOs.
+//        paio:  the completed AIOs
+//        paio->rval: the number of bytes read/written
+// or            zero if a timeout has occurred
+// or            any negative return will result in a assert upon return
+
+int aio_queue_t::get_next_completed(int fd, int timeout_ms, aio_t **paio, int max)
+{
+  struct timespec t = {
+    timeout_ms / 1000,
+    (timeout_ms % 1000) * 1000 * 1000
+  };
+
+  struct kevent events[max];
+  struct kevent ev;
+  dout(21) << __func__ << " kqueue registration"
+      << " on fd = " << fd
+      << " , with kqueue handle ctx = " << ctx
+      << dendl;
+
+  EV_SET(&ev, fd, 0, EV_ADD | EV_CLEAR, 0, 0, 0);
+  int rev = kevent(ctx, &ev, 1, NULL, 0, 0 );
+  if (rev < 0) {
```

Comment:
should also check for `ev.flags & EV_ERROR`.

Reply:
We are registering the event-filter. Resulting events are normally returned in the last three parameters, which are all unavailable because the size of the event-array is zero. 
So errors are returned by function-result and errno.

Later on you have the opposite, the filter is not changed (ev=NULL), and we are only asking for events to be delivered. (using the previously set filter)